File: android/src/main/kotlin/co/quis/flutter_contacts/crud/models/properties/Relation.kt
package co.quis.flutter_contacts.crud.models.properties

import android.content.ContentProviderOperation
import android.database.Cursor
import co.quis.flutter_contacts.common.CursorHelpers.getString
import co.quis.flutter_contacts.crud.models.JsonHelpers
import co.quis.flutter_contacts.crud.models.labels.Label
import co.quis.flutter_contacts.crud.models.labels.RelationLabel
import java.util.Objects
import android.provider.ContactsContract.CommonDataKinds.Relation as RelationData

data class Relation(
    val name: String,
    val label: Label = Label(label = "other"),
    val metadata: PropertyMetadata? = null,
) {
    companion object {
        fun fromJson(json: Map<String, Any?>): Relation =
            Relation(
                name = json["name"] as String,
                label = JsonHelpers.decodeRequiredObject(json, "label", Label::fromJson),
                metadata = JsonHelpers.decodeOptionalObject(json, "metadata", PropertyMetadata::fromJson),
            )

        fun fromCursor(cursor: Cursor): Relation =
            Relation(
                name = cursor.getString(RelationData.NAME),
                label = RelationLabel.fromCursor(cursor),
                metadata = PropertyHelpers.extractMetadata(cursor),
            )
    }

    fun toJson(): Map<String, Any?> {
        val result = mutableMapOf<String, Any?>("name" to name, "label" to label.toJson())
        JsonHelpers.encodeOptionalObject(result, "metadata", metadata) { it.toJson() }
        return result
    }

    fun toInsertOperation(
        rawContactId: Long? = null,
        rawContactIndex: Int = 0,
        addYield: Boolean = false,
    ): ContentProviderOperation {
        val labelPair = RelationLabel.toAndroidLabel(label.label, label.customLabel)
        return PropertyHelpers.buildInsertOperation(
            rawContactId,
            rawContactIndex,
            RelationData.CONTENT_ITEM_TYPE,
            addYield,
        ) {
            withValue(RelationData.NAME, name)
            withTypeAndLabel(RelationData.TYPE, RelationData.LABEL, labelPair)
        }
    }

    fun toUpdateOperation(): ContentProviderOperation {
        val labelPair = RelationLabel.toAndroidLabel(label.label, label.customLabel)
        return PropertyHelpers.buildUpdateOperation(
            PropertyHelpers.requireDataId(metadata, "relation"),
        ) {
            withValue(RelationData.NAME, name)
            withTypeAndLabel(RelationData.TYPE, RelationData.LABEL, labelPair)
        }
    }

    fun toDeleteOperation(): ContentProviderOperation =
        PropertyHelpers.buildDeleteOperation(
            PropertyHelpers.requireDataId(metadata, "relation"),
        )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Relation) return false
        return name == other.name && label == other.label
    }

    override fun hashCode() = Objects.hash(name, label)
}

File: lib/models/properties/relation.dart
import '../../utils/json_helpers.dart';
import '../labels/label.dart';
import '../labels/relation_label.dart';
import 'property_metadata.dart';

/// Relationship property.
class Relation {
  /// Name of the related person.
  final String name;

  /// Relationship label type.
  ///
  /// Defaults to [RelationLabel.other].
  final Label<RelationLabel> label;

  /// Property identity metadata (used internally for updates).
  final PropertyMetadata? metadata;

  const Relation({
    required this.name,
    Label<RelationLabel>? label,
    this.metadata,
  }) : label = label ?? const Label(RelationLabel.other);

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{'name': name};
    json['label'] = label.toJson();
    JsonHelpers.encode(json, 'metadata', metadata, (m) => m.toJson());
    return json;
  }

  static Relation fromJson(Map json) => Relation(
    name: json['name'] as String,
    label: Label.fromJson(
      json['label'] as Map,
      RelationLabel.values,
      RelationLabel.other,
    ),
    metadata: JsonHelpers.decode(json['metadata'], PropertyMetadata.fromJson),
  );

  @override
  String toString() => JsonHelpers.formatToString('Relation', toJson());

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      (other is Relation && name == other.name && label == other.label);

  @override
  int get hashCode => Object.hash(name, label);

  Relation copyWith({
    String? name,
    Label<RelationLabel>? label,
    PropertyMetadata? metadata,
  }) => Relation(
    name: name ?? this.name,
    label: label ?? this.label,
    metadata: metadata ?? this.metadata,
  );
}
